platform/search/src/GenericSearchProvider.js

```diff
@@ -0,0 +1,259 @@
+            // Takes the searchInput, as well as a max number of results (will return 
+            //   less than that if there are fewer matches).
+            function workerSearch(searchInput, maxResults, timestamp, timeout) {
+                var message = {
+                    request: 'search',
+                    input: searchInput,
+                    maxNumber: maxResults,
+                    timestamp: timestamp,
+                    timeout: timeout
+                };
+                worker.postMessage(message);
+            }
+            
+            // Handles responses from the web worker. Namely, the results of 
+            // a search request. 
+            function handleResponse(event) {
+                var ids = [],
+                    id;
+                
+                // If we have the results from a search 
+                if (event.data.request === 'search') {
+                    // Convert the ids given from the web worker into domain objects
+                    for (id in event.data.results) {
+                        ids.push(id);
+                    }
+                    objectService.getObjects(ids).then(function (objects) {
+                        var searchResults = [],
+                            id;
+                        
+                        // Create searchResult objects
+                        for (id in objects) {
+                            searchResults.push({
+                                object: objects[id],
+                                id: id,
+                                score: event.data.results[id]
+                            });
+                        }
+                        
+                        // Resove the promise corresponding to this 
+                        pendingQueries[event.data.timestamp].resolve({
+                            hits: searchResults,
+                            total: event.data.total,
+                            timedOut: event.data.timedOut
+                        });
+                    });
+                }
+            }
+            
+            worker.onmessage = handleResponse;
+            
+            // Helper function for getItems(). Indexes the tree.
+            function indexItems(nodes) {
+                nodes.forEach(function (node) {
+                    var id = node && node.getId && node.getId();
+                    
+                    // If we have not yet indexed this item, index it 
+                    if (!indexed[id]) {
+                        // Index each item with the web worker
+                        indexItem(node);
+                        indexed[id] = true;
+                        
+                        if (node.hasCapability && node.hasCapability('composition')) {
+                            // This node has children
+                            node.getCapability('composition').invoke().then(function (children) {
```

Comment:
First thing I'd try wrapping retrieval of composition in $q.when to ensure it is asynchronous:

`$q.when(node.getCapability('composition').invoke()).then(...`

Note that you can also shorthand the invoke

`$q.when(node.useCapability('composition')).then(...`

If this doesn't work, you could try using `$timeout` or `$interval` to split up the indexing operation over time (this might require a bit of refactoring, but basic idea is instead of immediately making the recursive call to `indexItems`, you'd do so on a timeout to break up the work over time.)